### PR TITLE
Web access management (WAM)

### PR DESCRIPTION
**Pull request by shapiromatron**
Enabled optional integration with Web Access Management (WAM). Currently only the admin side of the application requests authentication; there are no changes to the main page.  Note that this is largely a port of https://github.com/shapiromatron/hawc/pull/525. There are three options for authentication integration:

- With `AuthProvider.django`; no changes are made from the existing app [default]
- With `AuthProvider.external`; users cannot login or register using django views
- With both 1 and 2 enabled, both options are available on the admin login screen as shown:

image

The `get_user_metadata` method of `common.views.ExternalAuth` is implementation specific, so it needs to be overridden before use.

In addition to WAM options, the following additional changes were made:

A few additional changes:
- Added a new 401 error page for auth errors
- Updated documentation for configuration settings
- Added OpenAPI and swagger browser; both restricted to be admin only
- Moved urls to django project `main` since there are more views from multiple django applications
- Added `INCLUDE_ADMIN` option to include admin or disable; dynamically build django URL routes based on this
- Added `SESSION_DURATION`
- Added `LOGOUT_REDIRECT` option for logging out of shared authentication sessions

A few unrelated bugs fixed:
- fix disabled report cache
- fix `TEST_DB_FIXTURE` inside containers
- updated `sync.sh` in container for syncing database
- updated test structure to prevent unique filenaming

### CODE REVIEW

File: bmds_server/common/views.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                return HttpResponseRedirect(reverse("401"))
+        except User.DoesNotExist:
+            # Ensure email is unique
+            if User.objects.filter(email=email).exists():
+                self.mail_bad_auth(email, username)
+                return HttpResponseRedirect(reverse("401"))
+            # Create user
+            user = User.objects.create_user(email=email, username=username, **metadata)
+        login(request, user)
+        return HttpResponseRedirect(reverse("home"))
```
Comment (rabstejnek) [inline bmds_server/common/views.py:107]: Maybe redirect to admin?

File: bmds_server/common/views.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            if user.email != email:
+                self.mail_bad_auth(email, username)
+                return HttpResponseRedirect(reverse("401"))
+        except User.DoesNotExist:
+            # Ensure email is unique
+            if User.objects.filter(email=email).exists():
+                self.mail_bad_auth(email, username)
+                return HttpResponseRedirect(reverse("401"))
+            # Create user
+            user = User.objects.create_user(email=email, username=username, **metadata)
```
Comment (rabstejnek) [inline bmds_server/common/views.py:105]: Should we remove user create since we're dealing with an admin whitelist?

**Review by rabstejnek — APPROVED** [review]
Looks good! I just had two comments.

File: bmds_server/common/views.py (hunk 1 above)
Comment (shapiromatron) [reply bmds_server/common/views.py:107]: yes, good idea; will do

File: bmds_server/common/views.py (hunk 2 above)
Comment (shapiromatron) [reply bmds_server/common/views.py:105]: We will create the user here, but won't mark them as staff, that has to be explicitly granted by another admin.

With many external auth systems, there's authorization as well, so we won't get spurious users being created.

**Pull request merged**